### PR TITLE
spirv-val: Add Duplicate EntryPoint Builtin check

### PR DESCRIPTION
**Pull request by spencer-lunarg**
From https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/6625 (added in 1.3.285) (original issue https://github.com/KhronosGroup/SPIRV-Cross/issues/2313)

This adds 

- VUID-StandaloneSpirv-OpEntryPoint-09658
- VUID-StandaloneSpirv-OpEntryPoint-09659

image

### CODE REVIEW

File: source/val/validate_decorations.cpp [hunk 1]
```diff
@@ -848,9 +841,40 @@ spv_result_t CheckDecorationsOfEntryPoints(ValidationState_t& vstate) {
             break;
           if (auto error = CheckBuiltInVariable(interface, vstate))
             return error;
-        } else if (isBuiltInVar(interface, vstate)) {
-          if (auto error = CheckBuiltInVariable(interface, vstate))
-            return error;
+        } else {
```
Comment (alan-baker) [inline source/val/validate_decorations.cpp:844]: This misses the case that you have duplicates between the built-in block and loose variables. I expect the built-in block part of this if should also record which builtins are used appropriately.

File: source/val/validate_decorations.cpp [hunk 2]
```diff
@@ -848,9 +841,40 @@ spv_result_t CheckDecorationsOfEntryPoints(ValidationState_t& vstate) {
...
+          for (auto& dec : vstate.id_decorations(interface)) {
+            if (dec.dec_type() != spv::Decoration::BuiltIn) continue;
+
+            if (auto error = CheckBuiltInVariable(interface, vstate))
+              return error;
+
+            if (!spvIsVulkanEnv(vstate.context()->target_env)) continue;
+
+            const spv::BuiltIn builtin = dec.builtin();
+            if (storage_class == spv::StorageClass::Input) {
```
Comment (alan-baker) [inline source/val/validate_decorations.cpp:855]: Might make sense to refactor this a bit if get used in the built-in structure checks.

**Review by alan-baker — CHANGES_REQUESTED** [review]
Glslang unit tests seems to run a foul of this. Specifically some ray tracing tests where `gl_RayTMaxEXT` and `gl_HitTEXT` both generate the `RayTMaxKHR` built-in.

File: source/val/validate_decorations.cpp (hunk 1 above)
Comment (spencer-lunarg) [reply source/val/validate_decorations.cpp:844]: I thought the [whole idea](https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/6625#note_467693) of built-in block is the `builtin` is tied to the `OpTypeStruct` and not the `OpVariable`  so we don't need to check those

Comment (alan-baker) [reply source/val/validate_decorations.cpp:844]: The built-in block is still part of the interface and the point of that change is to say no builtin duplication in the interface. For a given interface, all the builtins may be on variables, but some might also be in a builtin block.

File: source/val/validate_decorations.cpp [hunk 3]
```diff
@@ -848,9 +841,43 @@ spv_result_t CheckDecorationsOfEntryPoints(ValidationState_t& vstate) {
...
+              // RayTmaxKHR can be duplicated
+              // (https://gitlab.khronos.org/vulkan/vulkan/-/issues/3885)
```
Comment (spencer-lunarg) [inline source/val/validate_decorations.cpp:857]: update - the glsl change we merged, so will remove this exception

File: source/val/validate_decorations.cpp (hunk 1 above)
Comment (spencer-lunarg) [reply source/val/validate_decorations.cpp]: re-worked, section of code, added tests as well

**Review by alan-baker — APPROVED** [review]

**Pull request merged**